[#140] Freeze/Unfreeze entrypoints (#164)
tqtezos/baseDAO: docs/specification-ligo.md

> ## Proposals

Reviewer: I suggest to add some info about the stages here and remove some details from below as a consequence:
```suggestion
## Stages

The contract constantly cycles between two stages, a proposing stage and a voting stage.
Both have the same same length, `voting_period` and alternate between each other,
starting from "voting" for period number `0`.

Tokens can be frozen in any period, but they can only be used for voting, proposing
and unfreezing starting from the following one and onwards.

### Proposals
```

Everyone can make a new proposal, however, you have to freeze some tokens for that.

> The proposer specifies how many frozen tokens they want to stake and this value is checked by
> the contract according to its compile-time configuration. Also tokens frozen in a voting period
> can only be used in the next period and further. This applies to proposal creation as well as to
> the `unfreeze` operation.

Reviewer: ```suggestion
The proposer specifies how many frozen tokens they want to stake and this value is checked by
the contract according to its compile-time configuration. 

This can only be performed in a proposing stage period, meaning one that's odd-numbered and
the proposer must have frozen his tokens in one of the preceding periods.
```

Proposals are identified by a key which is a `bytes` value computed via Blake2B hashing function of a
pair of propose entrypoint params and the proposer address.

> ## Voting

Reviewer: ```suggestion
### Voting
```

> Once a proposal is submitted, everyone can vote on it as long as they have enough frozen tokens to stake.
> One frozen token is required for one vote. Here also, only tokens frozen during the previous periods can
> be used to cast votes.

Reviewer: ```suggestion
Once a proposal is submitted, everyone can vote on it as long as they have enough frozen tokens to stake.
One frozen token is required for one vote. 

A vote can only be cast in a voting stage period, meaning one that's even-numbered.
Moreover the proposal vote on must have been submitted in the proposing period immediately preceeding
and the voter must have frozen his tokens in one of the preceding periods.
```

Voting period is specified for the whole smart contract and can be updated by the administrator; on update, the existing proposals are also affected.
It's possible to vote positively or negatively.

> After the voting ends, the contract is "flushed" by calling a dedicated entrypoint.

Reviewer: (writing this here because I cannot comment below)

3 new errors need to be added to the table below: `NOT_PROPOSING_PERIOD`, `NOT_ENOUGH_FROZEN_TOKENS`, and `NOT_ENOUGH_STAKED_TOKENS`

Author: Have added.

```

- The proposal is saved under `BLAKE2b` hash of proposal value and sender.
- The `Natural` value: `proposalTokenAmount` determines how many sender's frozen tokens will be staked in addition to the [fee](#set_fixed_fee_in_token).
- Sender MUST have enough frozen tokens (i. e. `≥ proposalTokenAmount + fee`) that are not already staked for a proposal or a vote.
- Fails with `NOT_ENOUGH_FROZEN_TOKENS` if the unstaked frozen token balance of the SENDER
  is less than `proposalTokenAmount + fee`.

> - Fails with `FAIL_PROPOSAL_CHECK` if the proposal is rejected by `proposalCheck` from the configuration.

Reviewer: Note: we need to add a line about the proposal stage requirement to the propose entrypoint description.

```suggestion
- Fails with `NOT_PROPOSING_PERIOD` if the current period is not a proposing stage.
- Fails with `FAIL_PROPOSAL_CHECK` if the proposal is rejected by `proposalCheck` from the configuration.
```

- This implements permits mechanism similar to the one in TZIP-17 but injected directly to the entrypoint.
- Vote author is identified by permit information, or if it is absent - `sender` is taken.
- Author MUST have unfrozen tokens equal to `voteAmount` or more (1 unstaked frozen token is needed for 1 vote) from past periods.
- Fails with `NOT_ENOUGH_FROZEN_TOKENS` if the frozen token balance of the author from past periods that is not staked
  is less than specified `voteAmount` .
- Fails with `PROPOSAL_NOT_EXIST` if the proposal key is not associated with any ongoing proposals.

> - Fails with `VOTING_PERIOD_OVER` if the proposal associated with the proposal key is already ended.

Reviewer: ```suggestion
- Fails with `VOTING_PERIOD_OVER` if the voting period of proposal associated with the proposal key has already ended.
```

- a `string`: the custom entrypoint name stored inside `custom_entrypoints` to execute
- a `bytes`: the `packed` representation of the `<ep_param>` to execute it with

### **freeze**

```ocaml

> type freeze_param = nat

Reviewer: ```suggestion
type freeze_param = nat

Freeze of freeze_param
```

```

> 

Reviewer: Need to add the Michelson representation too.

Author: Have added.

> Freezes the specified amount of tokens. The caller should have the required number of unfrozen tokens.
> Tokens frozen in a voting period can only be used from the next period onwards.
> 

Reviewer: ```suggestion
- Freezes the specified amount of tokens, which can only be used from the next period onward.
- Author MUST have unfrozen tokens equal to `freeze_param` or more.
```

### **unfreeze**

```ocaml

> type unfreeze_param = nat

Reviewer: ```suggestion
type unfreeze_param = nat

Unfreeze of unfreeze_param
```
Also need the Michelson representation as well.

```

> Unfreezes the specified amount of tokens. The caller should have the required number of frozen tokens
> that are not staked for a proposal or a vote. Tokens frozen in a voting period can only be unfrozen from
> the next period onwards.

Reviewer: ```suggestion
- unfreezes the specified amount of tokens from the tokens frozen in previous periods.
- Fails with `NOT_ENOUGH_FROZEN_TOKENS` if the author does not have enough tokens that can be unfrozen
```
